src/Components/Components/src/Routing/Router.cs

```diff
@@ -155,12 +190,51 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path)
+        {
+            // If this router instance does not provide an OnNavigateAsync parameter
+            // then we render the component associated with the route as per usual.
+            if (!OnNavigateAsync.HasDelegate)
+            {
+                return;
+            }
+
+            // If we've already invoked a task and stored its CTS, then
+            // cancel the existing task.
+            _onNavigateCts?.Dispose();
+
+            // Create a new cancellation token source for this instance
+            _onNavigateCts = new CancellationTokenSource();
+            var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
+            var task = OnNavigateAsync.InvokeAsync(navigateContext);
+
+            // Create a cancellation task based on the cancellation token
+            // associated with the current running task.
+            var cancellationTaskSource = new TaskCompletionSource();
+            navigateContext.CancellationToken.Register(() =>
+                cancellationTaskSource.SetResult());
```

Comment:
We can avoid a closure here:

```c#
            navigateContext.CancellationToken.Register(state =>
                ((TaskCompletionSource)state).SetResult(), cancellationTaskSource);
```

Also, nit: I think `cancellationTaskSource` should be `cancellationTcs`. I thought I was looking at a CTS at first.

Reply:
Also this probably needs to move to before the OnNavigateAsync

Reply:
Good call. Pranav made a similar suggestion back when this was doing state.TrySetCancelled.